ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -349,6 +349,13 @@ let assert_host_has_iommu ~__context ~host =
   if List.assoc "iommu" chipset_info <> "true" then
     raise (Api_errors.Server_error (Api_errors.vm_requires_iommu, [Ref.string_of host]))
 
+(* Check if there are vgpus not allocated, since we may check the mem/vgpu after reservation *)
+let has_non_allocated_vgpus ~__context ~self =
+  let resident_pgpus = Db.VM.get_VGPUs ~__context ~self
+                       |> List.map (fun vgpu -> Db.VGPU.get_scheduled_to_be_resident_on ~__context ~self:vgpu) in
+  let non_allocated = List.filter (fun pgpu -> not (Db.is_valid_ref __context pgpu)) resident_pgpus in
+  non_allocated <> []
```

Comment:
You can make this code a long chain:
```ocaml
Db.VM.get_VGPUs ~__context ~self
|> List.map (fun vgpu -> Db.VGPU.get_scheduled_to_be_resident_on ~__context ~self:vgpu)
|> List.filter (fun pgpu -> not (Db.is_valid_ref __context pgpu))
|> (<>) []
```

Or even
```ocaml
Db.VM.get_VGPUs ~__context ~self
|> List.exists (fun vgpu -> not (
     Db.VGPU.get_scheduled_to_be_resident_on ~__context ~self:vgpu
     |> Db.is_valid_ref __context
))
```

Reply:
Thanks, it really looks good. I'd like to change to the first one. Thanks. @mseri